executor/studentapi.pyx

```diff
@@ -0,0 +1,272 @@
+        self.context = context
+
+    def __repr__(self):
+        cls_name, (msg, *_) = self.__class__.__name__, self.args
+        kwargs = ', '.join(f'{name}={repr(value)}' for name, value in self.context.items())
+        return f'{cls_name}({repr(msg)} {kwargs})'
+
+
+cdef class Gamepad:
+    """
+    The API for accessing gamepads.
+
+    Attributes:
+        mode: The execution state of the robot.
+    """
+    cdef public str mode
+
+    def __cinit__(self, mode):
+        """Initializes the mode of the robot. Also initializes the auxiliary SHM. """
+        self.mode = mode
+        # shm_aux_init(STUDENTAPI)
+        # log_runtime(DEBUG, "Aux SHM initialized")
+
+    def __dealloc__(self):
+        """Once process is finished and object is deallocated, close the mapping to the auxiliary SHM."""
+        # shm_aux_stop(STUDENTAPI)
+        # log_runtime(DEBUG, "Aux SHM stopped")
+
+    cpdef get_value(self, str param_name):
+        """
+        Get a gamepad parameter if the robot is in teleop.
+
+        Args:
+            param: The name of the parameter to read. Possible values are at https://pioneers.berkeley.edu/software/robot_api.html 
+        """
+        if self.mode != 'teleop':
+            raise CodeExecutionError(f'Cannot use Gamepad during {self.mode}')
+        # Convert Python string to C string
+        cdef bytes param = param_name.encode('utf-8')
+        cdef uint32_t buttons
+        cdef float joysticks[4]
+        gamepad_read(&buttons, joysticks)
+        cdef char** button_names = get_button_names()
+        cdef char** joystick_names = get_joystick_names()
+        # Check if param is button
+        for i in range(NUM_GAMEPAD_BUTTONS):
+            if param == button_names[i]:
+                return bool(buttons & (1 << i))
+        # Check if param is joystick
+        for i in range(4):
+            if param == joystick_names[i]:
+                return joysticks[i]
+        raise KeyError(f"Invalid gamepad parameter {param_name}")
+
+
+class ThreadWrapper(threading.Thread):
+
+    def __init__(self, action, args, kwargs):
+        super().__init__()
+        self.action = action
+        self.args = args
+        self.kwargs = kwargs
+
+    def run(self):
+        try:
+            self.action(*self.args, **self.kwargs)
+        except TimeoutError:
+            pass
+        except Exception as e:
+            _print(str(e), level=ERROR)
+
+
+cdef class Robot:
+    """
+    The API for accessing the robot and its devices.
+    """
+    cdef public dict running_actions
+
+    def __cinit__(self):
+        """Initializes the shared memory (SHM) wrapper. """
+        # shm_init(STUDENTAPI) # Remove once integrated into executor?
+        # log_runtime(DEBUG, "SHM intialized")
+        self.running_actions = {}
+
+    def __dealloc__(self):
+        """Once process is done and object is deallocated, close the mapping to SHM."""
+        # shm_stop(STUDENTAPI)
+        # log_runtime(DEBUG, "SHM stopped")
+
```

Comment:
Same comment here as with the Gamepad class~~

Reply:
Removed